libglusterfs/src/iobuf.c

```diff
@@ -468,79 +441,38 @@ __iobuf_get(struct iobuf_pool *iobuf_pool, const size_t page_size,
 }
 
 static struct iobuf *
-iobuf_get_from_stdalloc(struct iobuf_pool *iobuf_pool, const size_t page_size)
+iobuf_get_from_stdalloc(const size_t page_size, gf_boolean_t align_flag)
 {
     struct iobuf *iobuf = NULL;
-    struct iobuf_arena *iobuf_arena = NULL;
-    struct iobuf_arena *trav = NULL;
-    int ret = -1;
+    size_t align_size = page_size;
 
-    /* The first arena in the 'MAX-INDEX' will always be used for misc */
-    list_for_each_entry(trav, &iobuf_pool->arenas[IOBUF_ARENA_MAX_INDEX], list)
-    {
-        iobuf_arena = trav;
-        break;
-    }
-
-    iobuf = GF_CALLOC(1, sizeof(*iobuf), gf_common_mt_iobuf);
+    iobuf = GF_MALLOC(sizeof(*iobuf), gf_common_mt_iobuf);
     if (!iobuf)
         goto out;
 
-    /* 4096 is the alignment */
-    iobuf->free_ptr = GF_CALLOC(1, ((page_size + GF_IOBUF_ALIGN_SIZE) - 1),
-                                gf_common_mt_char);
-    if (!iobuf->free_ptr)
-        goto out;
-
-    iobuf->ptr = GF_ALIGN_BUF(iobuf->free_ptr, GF_IOBUF_ALIGN_SIZE);
-    iobuf->iobuf_arena = iobuf_arena;
-    iobuf->page_size = page_size;
-    LOCK_INIT(&iobuf->lock);
+    if (align_flag)
+        align_size = page_size + GF_IOBUF_ALIGN_SIZE;
```

Comment:
```suggestion
        align_size = page_size + GF_IOBUF_ALIGN_SIZE - 1;
```

Otherwise, in cases that `page_size` is already aligned to `GF_IOBUF_ALIGN_SIZE`, it will waste one entire page of memory.

Reply:
Certainly one of the properties we've lost in this change is that iobuf->ptr are no longer naturally aligned and we need now more memory than we've used to. This is noticeable in all block sizes:
- When in the past a 4K iobuf would have been just 4K taken from a mmap'ed arena, now we need 8K
- A 128K (which is our default page size, so all iobuf_get() use it) that could have been perfectly aligned as it was taken from an mmap'ed arena, now requires more, so doesn't really fit into tcmalloc's 128K recommendation.

The first means that we can't easily remove (for example) the extra memory copies in __posix_write(), the 2nd means that ```if (page_size <= USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {``` is probably not optimal. A hack would be convert:
```
if (page_size <= USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {
```
to:
```
if (page_size < USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {
```